src/core/ext/filters/client_channel/resolver/dns/event_engine/event_engine_client_channel_resolver.cc

```diff
@@ -0,0 +1,535 @@
+      resolver_->name_to_resolve(), grpc_core::kDefaultSecurePort,
+      resolver_->query_timeout_ms_);
+  GRPC_EVENT_ENGINE_DNS_TRACE(
+      "DNSResolver::%p Started resolving hostname. Handle::%s", resolver_.get(),
+      HandleToString(*hostname_handle_).c_str());
+  if (resolver_->enable_srv_queries_) {
+    srv_handle_ = event_engine_resolver_->LookupSRV(
+        absl::bind_front(&EventEngineDNSRequestWrapper::OnSRVResolved,
+                         Ref(DEBUG_LOCATION, "OnSRVResolved"),
+                         resolver_->name_to_resolve()),
+        resolver_->name_to_resolve(), resolver_->query_timeout_ms_);
+    GRPC_EVENT_ENGINE_DNS_TRACE(
+        "DNSResolver::%p Started resolving SRV. Handle::%s", resolver_.get(),
+        HandleToString(*srv_handle_).c_str());
+  }
+  if (resolver_->request_service_config_) {
+    txt_handle_ = event_engine_resolver_->LookupTXT(
+        absl::bind_front(&EventEngineDNSRequestWrapper::OnTXTResolved,
+                         Ref(DEBUG_LOCATION, "OnTXTResolved")),
+        resolver_->name_to_resolve(), resolver_->query_timeout_ms_);
+    GRPC_EVENT_ENGINE_DNS_TRACE(
+        "DNSResolver::%p Started resolving TXT. Handle::%s", resolver_.get(),
+        HandleToString(*txt_handle_).c_str());
+  }
+}
+
+EventEngineDNSRequestWrapper::~EventEngineDNSRequestWrapper() {
+  resolver_.reset(DEBUG_LOCATION, "dns-resolving");
+}
+
+void EventEngineDNSRequestWrapper::Orphan() {
+  {
+    MutexLock lock(&on_resolved_mu_);
+    orphaned_ = true;
+    // Event if cancellation fails here, OnResolvedLocked will return early, and
+    // the resolver will never see a completed request.
+    if (hostname_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*hostname_handle_);
+    }
+    if (srv_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*srv_handle_);
+    }
+    for (const auto& handle : balancer_hostname_handles_) {
+      event_engine_resolver_->CancelLookup(handle);
+    }
+    if (txt_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*txt_handle_);
+    }
+  }
+  Unref(DEBUG_LOCATION, "Orphan");
+}
+
+void EventEngineDNSRequestWrapper::OnHostnameResolved(
+    absl::StatusOr<std::vector<EventEngine::ResolvedAddress>> addresses) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    hostname_handle_.reset();
+    addresses_ = addresses;
+    result = OnResolvedLocked();
+  }
+  if (result.has_value()) {
+    resolver_->OnRequestComplete(std::move(*result));
+  }
+}
+
+void EventEngineDNSRequestWrapper::OnSRVResolved(
+    std::string srv_host,
+    absl::StatusOr<std::vector<EventEngine::DNSResolver::SRVRecord>>
+        srv_records) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    srv_handle_.reset();
+    if (srv_records.ok()) {
```

Comment:
Done. OnBalancerHostnamesResolved is a bit different since OnResolvedLocked is called in both success and failure scenarios, but it's still cleaner with abseil's cleanup.